test/upgrade-tests.sh

```diff
@@ -0,0 +1,152 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script runs the end-to-end tests against Knative Serving built from source.
+# It is started by prow for each PR. For convenience, it can also be executed manually.
+
+# If you already have the *_OVERRIDE environment variables set, call
+# this script with the --run-tests arguments and it will start knative in
+# the cluster and run the tests.
+
+# Calling this script without arguments will create a new cluster in
+# project $PROJECT_ID, start knative in it, run the tests and delete the
+# cluster.
+
+source $(dirname $0)/cluster.sh
+
+# Latest serving release. This is intentionally hardcoded so that we can test
+# upgrade/downgrade on release branches (or even arbitrary commits).
+#
+# Unfortunately, that means we'll need to manually bump this version when we
+# make new releases.
+#
+# Fortunately, that's not *too* terrible, because forgetting to bump this
+# version will make tests either:
+# 1. Still pass, meaning we can upgrade from earlier than latest release (good).
+# 2. Fail, which might be remedied by bumping this version.
+readonly SERVING_RELEASE_YAML=https://github.com/knative/serving/releases/download/v0.2.0/release-no-mon.yaml
+
+function create_serving_latest_release() {
+  echo ">> Bringing up Serving"
+  kubectl apply -f ${SERVING_RELEASE_YAML}
+  # Due to the lack of Status in Istio, we have to ignore failures in initial requests.
+  #
+  # However, since network configurations may reach different ingress pods at slightly
+  # different time, even ignoring failures for initial requests won't ensure subsequent
+  # requests will succeed all the time.  We are disabling ingress pod autoscaling here
+  # to avoid having too much flakes in the tests.  That would allow us to be stricter
+  # when checking non-probe requests to discover other routing issues.
+  #
+  # We should revisit this when Istio API exposes a Status that we can rely on.
+  # TODO(tcnghia): remove this when https://github.com/istio/istio/issues/882 is fixed.
+  echo ">> Patching Istio"
+  kubectl patch hpa -n istio-system knative-ingressgateway --patch '{"spec": {"maxReplicas": 1}}'
+}
+
+function create_everything_latest_release() {
+  # We assume the same istio install as HEAD. Dealing with istio upgrades is
+  # outside of the scope of this test; though this test should catch issues
+  # caused by incompatible istio versions.
+  create_istio
+
+  create_serving_latest_release
+}
+
+function delete_serving_latest_release() {
+  echo ">> Bringing down Serving"
+  kubectl delete --ignore-not-found=true -f ${SERVING_RELEASE_YAML}
+}
+
+# Cribbed from e2e-tests.sh
+function publish_test_images() {
```

Comment:
After #2320 is in, this function will be available in `cluster.sh`.

Reply:
Added a TODO.